<redacted> <redacted>
Repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
Commit ID: 7859840bb537a1740a64cf852f075fd740bbf328

In [1]:
# Imports section
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from IPython.display import display, Markdown
from sklearn.model_selection import KFold
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('./science_data_large.csv')

print(data.head(16))

print("Summary of the dataset:")
print(data.info())

print("Statistical summary:")
print(data.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
15              62       293  3.952883e+04
Summary of the dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X = data[['Temperature °C','Mols KCL']]
y = data['Size nm^3']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()
model.fit(X_train, y_train)

sample_datapoint = [[400, 200]]  
predicted_size = model.predict(sample_datapoint)
print(f"Predicted Size for KCl={sample_datapoint[0][0]}, Temperature={sample_datapoint[0][1]}: {predicted_size[0]:.2f}")

score = model.score(X_test, y_test)
print(f"Model R² score on test set: {score:.2f}")

coefficients = model.coef_
intercept = model.intercept_
print(f"Coefficients: {coefficients}")



Predicted Size for KCl=400, Temperature=200: 143606.10
Model R² score on test set: 0.86
Coefficients: [ 866.14641337 1032.69506649]


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


For the sample data of KCI = 400 and Temperature = 200, the model we trained predicts that the size will be 143606.10 nm^3.

$h(x) = -409391.47958340764 + 866.14641337 \times \text{KCl} + 1032.69506649 \times \text{Temperature}$



## Part 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

from sklearn.model_selection import KFold

cv_scores = cross_val_score(model, X, y, cv=5)
print(f"Cross-validation scores: {cv_scores}")
print(f"Mean cross-validation score: {cv_scores.mean():.2f}")

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.86


Cross validation is performed by splitting the data into K subsets, training the model on K - 1 subsets, and evaluating the model on the remaining subset. The mean cross-validation score of 0.86 tells us that the model performed similarly well on different subsets, and that our model has strong predictive power. 

## Part 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.

degree = 2
poly_model = make_pipeline(PolynomialFeatures(degree), LinearRegression())

poly_model.fit(X_train, y_train)

poly_pred = poly_model.predict(X_test)
poly_score = poly_model.score(X_test, y_test)
print(f"Polynomial Regression R² score on test set: {poly_score:.2f}")

poly_coefficients = poly_model.named_steps['linearregression'].coef_
print(poly_coefficients)
poly_intercept = poly_model.named_steps['linearregression'].intercept_

Polynomial Regression R² score on test set: 1.00
[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]


$h(x) = 0 + 12.00 \times \text{KCl} - 0.00000013 \times \text{Temperature} + 0.0000000000126 \times \text{KCl}^2 + 2.00 \times \text{Temperature}^2 + 0.02857 \times (\text{KCl} \times \text{Temperature})$
The $R^2$ term is 1.00 which would indicate a perfect fit.
